thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detection/algorithm/stage/AnomalyDetectionStageWrapper.java

```diff
@@ -67,4 +119,86 @@ public DetectionPipelineResult run() {
   private AnomalyDetectionStage loadAnomalyDetectorStage(String className) throws Exception {
     return (AnomalyDetectionStage) Class.forName(className).newInstance();
   }
+
+  List<Interval> getMonitoringWindows() {
+    if (this.movingWindowDetection) {
+      try{
+        List<Interval> monitoringWindows = new ArrayList<>();
+        MetricEntity me = MetricEntity.fromURN(this.metricUrn);
+        MetricConfigDTO metricConfigDTO =
+            this.provider.fetchMetrics(Collections.singletonList(me.getId())).get(me.getId());
+        dataset = this.provider.fetchDatasets(Collections.singletonList(metricConfigDTO.getDataset()))
+            .get(metricConfigDTO.getDataset());
+        dateTimeZone = DateTimeZone.forID(dataset.getTimezone());
+        List<Long> monitoringWindowEndTimes = getMonitoringWindowEndTimes();
+        for (long monitoringEndTime : monitoringWindowEndTimes) {
+          long endTime = monitoringEndTime - TimeUnit.MILLISECONDS.convert(windowDelay, windowDelayUnit);
+          long startTime = endTime - TimeUnit.MILLISECONDS.convert(windowSize, windowUnit);
+          monitoringWindows.add(new Interval(startTime, endTime));
+        }
+        return monitoringWindows;
+      } catch (Exception e) {
+        LOG.info("can't generate moving monitoring windows, calling with single detection window", e);
+      }
+    }
+    return Collections.singletonList(new Interval(startTime, endTime));
+  }
+
+  private List<Long> getMonitoringWindowEndTimes() {
+    List<Long> endTimes = new ArrayList<>();
+
+    // get current hour/day, depending on granularity of dataset,
+    DateTime currentEndTime = new DateTime(getBoundaryAlignedTimeForDataset(new DateTime(endTime, dateTimeZone)), dateTimeZone);
+
+    DateTime lastDateTime = new DateTime(getBoundaryAlignedTimeForDataset(new DateTime(startTime, dateTimeZone)), dateTimeZone);
+    Period bucketSizePeriod = getBucketSizePeriodForDataset();
+    while (lastDateTime.isBefore(currentEndTime)) {
```

Comment:
isBefore does a less than check right? Won't currentEndTime be excluded then?

Reply:
yes. this is the expected behavior as we do currently.